### ISSUE TITLE
I2S and DMA: Trouble getting Circular to work, and what are differences between Normal Circular vs. Link Circular?

### ISSUE THREAD

**Issue by nyh-workshop** [IOsetting/wm-sdk-w806#23]
Hi there,
I am trying to write a simple I2S and DMA example of a circular buffer with only the DMA TX.
However, I could not get these "Half Complete" and "Complete" interrupts to be triggered. There is no response after calling HAL_I2S_Transmit_DMA. All it does now is to jump to the while loop in the main.

In the library, there are two DMA transfer modes: Link and Normal. I am not entirely sure what are the differences between that, so I picked the Normal. 

Here are the parts of the code, which I modified from the i2s example:

**main.c**

```
int main(void)
{
    SystemClock_Config(CPU_CLK_160M);
    printf("enter main\r\n");
    
    HAL_Init();
    DMA_Init();
    I2S_Init();
    
    // I2S_BUFFER_SIZE is 512 in my modified example:
    memset(tx_buf, 0x00, I2S_BUFFER_SIZE);

    HAL_I2S_Transmit_DMA(&hi2s, (uint32_t*)tx_buf, I2S_BUFFER_SIZE);

    while (1)
    {
        printf(".");
        HAL_Delay(1000);
    }
}

static void DMA_Init(void)
{
    __HAL_RCC_DMA_CLK_ENABLE();
    
    HAL_NVIC_SetPriority(DMA_Channel0_IRQn, 0);
    HAL_NVIC_EnableIRQ(DMA_Channel0_IRQn);
    
}

void HAL_I2S_TxHalfCpltCallback(I2S_HandleTypeDef *hi2s)
{
    printf("tx halfcplt\r\n");
}

void HAL_I2S_TxCpltCallback(I2S_HandleTypeDef *hi2s)
{
    printf("tx cplt\r\n");
}
```

**wm_hal_msp.c:**
```
void HAL_I2S_MspInit(I2S_HandleTypeDef *hi2s)
{
    if (hi2s->Instance == I2S)
    {
        __HAL_RCC_I2S_CLK_ENABLE();
        __HAL_RCC_GPIO_CLK_ENABLE();
        
        __HAL_AFIO_REMAP_I2S_MCK(GPIOA, GPIO_PIN_7);
        __HAL_AFIO_REMAP_I2S_WS(GPIOB, GPIO_PIN_9);
        __HAL_AFIO_REMAP_I2S_CK(GPIOB, GPIO_PIN_8);
        __HAL_AFIO_REMAP_I2S_MOSI(GPIOB, GPIO_PIN_11);
        __HAL_AFIO_REMAP_I2S_MISO(GPIOB, GPIO_PIN_10);
        
        hdma_i2s_tx.Instance = DMA_Channel0;
        hdma_i2s_tx.Init.Direction = DMA_MEMORY_TO_PERIPH;
        hdma_i2s_tx.Init.DestInc = DMA_DINC_DISABLE;
        hdma_i2s_tx.Init.SrcInc = DMA_SINC_CIRCULAR;
        hdma_i2s_tx.Init.DataAlignment = DMA_DATAALIGN_WORD;
        hdma_i2s_tx.Init.Mode = DMA_MODE_NORMAL_CIRCULAR;
        hdma_i2s_tx.Init.RequestSourceSel = DMA_REQUEST_SOURCE_I2S_TX;
        
        hdma_i2s_tx.LinkDesc = tx_desc;
        
        if (HAL_DMA_Init(&hdma_i2s_tx) != HAL_OK)
        {
            Error_Handler();
        }
        __HAL_LINKDMA(hi2s, hdmatx, hdma_i2s_tx);   

        HAL_NVIC_SetPriority(I2S_IRQn, 1);
        HAL_NVIC_EnableIRQ(I2S_IRQn);
    }
}
```

Is the Circular mode, when running, should trigger both the Half-Transfer Complete and Transfer Complete interrupts continuously?

**Comment by IOsetting** [comment]
Apologies I didn't try I2S and DMA myself so I am not quite sure about the problem.

```c
uint32_t Mode;                  /*  Specifies the operation mode of the DMA channelx. When you need to 
                                        use half-complete interrupts, you must use link mode.
                                        This parameter can be a value of @ref DMA_MODE*/
```
**link mode** means two arrays in DMA mode, when transmitting one array the data will be filled into another, something like double-buffer, to increase performance. 

**Half Complete** works with **link mode**, the interrupt will be triggered when one array has finished transmitting and the software(code) need to set the flag to notify DMA that another array is ready.

This blog may provide more information about DMA https://blog.csdn.net/weixin_44457994/article/details/122711562 (can be translated to English in Google Chrome)

**Comment by nyh-workshop** [comment]
Hello, thanks for the reply.

I have examined the I2S and DMA code and noticed something when I tried to experiment with it.

Since the HAL is possibly based on the STM32 ones, I looked the examples and found out that the "Transfer Complete" callback has been disabled in the code (wm_i2s.c):

```
static void I2S_DMATxCplt(DMA_HandleTypeDef *hdma)
{
    I2S_HandleTypeDef *hi2s = (I2S_HandleTypeDef *)((DMA_HandleTypeDef *)hdma)->Parent;
    
    if ((hdma->Init.Mode == DMA_MODE_NORMAL_SINGLE) || (hdma->Init.Mode == DMA_MODE_LINK_SINGLE))
    {
        CLEAR_BIT(hi2s->Instance->CR, I2S_CR_TXDMA_EN);
    //    __HAL_I2S_DISABLE_TX(hi2s);
        hi2s->TxXferCount = 0;
    //    hi2s->State = HAL_I2S_STATE_READY;
    }
//    HAL_I2S_TxCpltCallback(hi2s); // <- this callback has been disabled in the code.
}
```

I'm not sure if there are reasons this is disabled in the driver code.

As I'm curious on this, I have to manually enable it (by removing that comment) and now it gives a "Half Complete" and "Transfer Complete" interrupts at **Linked Circular** mode. The I2S and DMA transfer is now being performed automatically with very little user intervention:

![image](https://user-images.githubusercontent.com/20377029/180654945-a5c8dffb-56d8-4605-898b-c46dea35c29b.png)

Meanwhile I'm still experimenting with the rest of the double-buffering code. If this code works well, is it alright for me to send in my sine wave I2S example?
